File: ddpoliglot_v6_flutter/lib/models/data/WordPhraseTranslation.dart
// ignore_for_file: file_names, hash_and_equals, annotate_overrides
import '../player_classes.dart';
import './Language.dart';
import './WordPhrase.dart';

import 'package:json_annotation/json_annotation.dart';

part 'WordPhraseTranslation.g.dart';

@JsonSerializable()
class WordPhraseTranslation extends Plailable {
  int wordPhraseTranslationID;
  int wordPhraseID;
  WordPhrase? wordPhrase;
  int languageID;
  Language? language;
  String text;
  int hashCode;
  String? textSpeechFileName;
  int speachDuration;

  WordPhraseTranslation(
    this.wordPhraseTranslationID,
    this.wordPhraseID,
    this.wordPhrase,
    this.languageID,
    this.language,
    this.text,
    this.hashCode,
    this.textSpeechFileName,
    this.speachDuration,
  );

  factory WordPhraseTranslation.fromJson(Map<String, dynamic> data) =>
      _$WordPhraseTranslationFromJson(data);

  Map<String, dynamic> toJson() => _$WordPhraseTranslationToJson(this);
}

File: ddPoliglotV6.Data/Models/WordPhraseTranslation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;

namespace ddPoliglotV6.Data.Models
{
    public class WordPhraseTranslation
    {
        [Key]
        public int WordPhraseTranslationID { get; set; }

        public int WordPhraseID { get; set; }

        [ForeignKey("WordPhraseID")]
        public WordPhrase WordPhrase { get; set; }

        public int LanguageID { get; set; }

        [ForeignKey("LanguageID")]
        public Language Language { get; set; }

        [MaxLength(250)]
        public string Text { get; set; }

        public Int64 HashCode { get; set; }

        [MaxLength(100)]
        public string TextSpeechFileName { get; set; }

        public int SpeachDuration { get; set; }

        public object Clone()
        {
            return this.MemberwiseClone();
        }

        public string Compare(WordPhraseTranslation other)
        {
            string[] propNames = new string[] {
                "LanguageID",
                "Text",
                "HashCode",
                "TextSpeechFileName",
                "SpeachDuration"
            };

            var result = new StringBuilder();
            foreach (var propName in propNames)
            {
                CompareProp(this, other, propName, result);
            }

            return result.ToString();
        }

        private void CompareProp(object curr, object other, string propName, StringBuilder result)
        {
            var currValue = curr.GetType().GetProperty(propName).GetValue(curr, null);
            var otherValue = other.GetType().GetProperty(propName).GetValue(other, null);
            if ((currValue?.ToString() ?? "null") != (otherValue?.ToString() ?? "null"))
            {
                result.AppendLine($"{propName}: {currValue} -> {otherValue}");
            }
        }

    }
}
